docs/add-code-flow.md

```diff
@@ -0,0 +1,97 @@
+
+#### `[MFS] PutNode(mfs.Root, path, ipld.Node)`
+*Insert node at path into given `MFS`*
+
+https://github.com/ipfs/go-mfs/blob/master/ops.go#L101-L113
+
+In this the path is used to determine the `MFS` `Directory`, which is first lookup up in the `MFS` using `lookupDir()` function. This is followed by adding the **root** __DAG__ node (`ipld.Node`) in to the found `Directory` using `directory.AddChild()` method.
+
+#### - `directory.AddChild(filename, ipld.Node)`
+*Add **root** __DAG__ node , as filename, under this directory*
+
+https://github.com/ipfs/go-mfs/blob/master/dir.go#L381-L402
+
+Within this method the node is added to the __DAG service__ of the `Directory` object using the `dserv.Add()` method [HELP NEEDED].
+This is subsequently followed by adding the **root** __DAG__ node by creating a `directory.child{}` object with the given name, within in the `directory.addUnixFSChild(directory.child{name, ipld.Node})` method.
+
+#### -- `directory.addUnixFSChild(child)`
+*Switch to HAMT (if configured) and add child to inner UnixFS Directory*
+
+https://github.com/ipfs/go-mfs/blob/master/dir.go#L406-L425
+
+Here the transition of the UnixFS directory to __HAMT__ implemetation is done, if configured, wherein if the directory is of type `BasicDirectory`, it is converted to a __HAMT__ implementation.
+The node is then added as a child to the inner `UnixFS` directory using the `directory.AddChild()` method.
+Note: This is not to be confused with the `directory.AddChild(filename, ipld.Node)`, as this operates on the inner `UnixFS` `Directory` object only.
+
+#### --- (inner)`Directory.AddChild(ctx, name, ipld.Node)`
+
+This method vastly differs based on the inner `Directory` implementation (Basic vs HAMT). Let's focus on the `BasicDirectory` implementation to keep things simple.
+
+https://github.com/ipfs/go-unixfs/blob/master/io/directory.go#L142-L147
+
+> IMPORTANT
+> It should be noted that the inner `Directory` of the `UnixFS` package, encasulates a node object of type `ProtoNode`, which is a different format as compared to the `ipld.Node` we have been working on throughout this document.
+
+This method first attempts to remove any old links (`ProtoNode.RemoveNodeLink(name)`) to the `ProtoNode` prior to adding a link to the newly added `ipld/Node`, using `ProtoNode.AddNodeLink(name, ipld.Node)`.
+
+https://github.com/ipfs/go-merkledag/blob/master/node.go#L99-L112
+
+The `AddNodeLink()` method is where an `ipld.Link` is created with the `ipld.Node`'s `CID` and size in the `ipld.MakeLink(ipld.Node)` method, and is then appended to the `ProtoNode`'s links in the `ProtoNode.AddRawLink(name)` method.
+
+---
+
+## `adder.Finalize()`
+*Fetch and return the __DAG__ **root** from the `MFS` and `UnixFS` directory*
+
+https://github.com/ipfs/go-ipfs/blob/master/core/coreunix/add.go#L199-L244
+
+The whole process ends with `adder.Finalize()` which returns the `ipld.Node` from the `UnixFS` `Directory`.
+**(HELP: Do we need to elaborate ?)**
```

Comment:
I don't think so, we should just mention that at the end of `AddAllAndPin` (after `Finalize`) we call `PinRoot` to recursively pin all the files under the MFS root.